### PR TITLE
[Provisioner] Avoid backward compatibility issue with provisioner

### PR DESCRIPTION
**Pull request by Michaelvll**
Previously, the cloud-specific provisioner will only be imported when the `sky.provision.<cloud>` is used for termination and stop for spot jobs (when provisioning logic is not implemented for the cloud). It will cause backward compatibility issue, when the following happens:
1. `sky spot launch` on master (job id: 1)
2. a new commit updated `sky.provision.<cloud>`
3. `sky spot launch` another job, which updates the skypilot library on the spot controller
4. `sky spot cancel -a`, which will load the latest `sky.provision.<cloud>`, which may not be compatible with other old code the spot job 1 used.

In this PR, we import the provision modules in advance to avoid the backward compatibility issue for any changes in `sky.provision` and `sky.adaptors.<cloud>` after this PR.

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky launch --cloud aws --cpus 2 -i 0 echo hi` with only aws dependency installed.
  - [x] `sky launch --cloud gcp --cpus 2 -i 0 echo hi; sky status -r; sky down cluster-name` with only gcp dependency installed.
  - [x] `sky launch --cloud azure --cpus 2 -i 0 echo hi` with only azure dependency installed.
- [x] All smoke tests: `pytest tests/test_smoke.py --aws` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [x] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh` for aws
- [x] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh` for gcp

### CODE REVIEW

File: sky/provision/__init__.py [hunk 1]
```diff
@@ -4,13 +4,15 @@
...
-import importlib
 import inspect
 from typing import Any, Dict, List, Optional
 
 from sky import sky_logging
 from sky import status_lib
+from sky.provision import aws
+from sky.provision import azure
 from sky.provision import common
+from sky.provision import gcp
```
Comment (concretevitamin) [inline sky/provision/__init__.py:15]: Should add a comment on: ...these imports should never fail even if underlying cloud SDK dependencies are not installed. This is ensured by using the .... pattern inside these modules.

File: sky/provision/__init__.py [hunk 2]
```diff
@@ -27,8 +29,9 @@ def _wrapper(*args, **kwargs):
         else:
             provider_name = kwargs.pop('provider_name')
 
-        module_name = provider_name
-        module = importlib.import_module(f'sky.provision.{module_name.lower()}')
+        module_name = provider_name.lower()
+        module = globals().get(module_name)
+        assert module is not None, f'Unknown provider: {module_name}'
```
Comment (concretevitamin) [inline sky/provision/__init__.py:34]: Should it be assert or raise RuntimeError?

**Review by concretevitamin — APPROVED** [review]
LGTM, may need this in soon as a hot fix. @suquark should take a look asap/after?

File: sky/provision/__init__.py [hunk 3]
```diff
@@ -4,13 +4,18 @@
...
 
 from sky import sky_logging
 from sky import status_lib
+# These provision.<cloud> modules should never fail even if underlying cloud SDK
+# dependencies are not installed. This is ensured by using sky.adaptors inside
+# these modules, for lazy loading of cloud SDKs.
+from sky.provision import aws
+from sky.provision import azure
 from sky.provision import common
+from sky.provision import gcp
```
Comment (suquark) [inline sky/provision/__init__.py:18]: later we may scan over the directory to import all modules automatically. I am fine with the current implementation though.

**Review by suquark — COMMENTED** [review]
LGTM

**Pull request merged**